### PR TITLE
Integrating CORS

### PR DESCRIPTION
**Pull request by UtkarshVerma**
Implemented the usage of Cross-Origin Resource Sharing. All the settings can be modified by changing the values of `cors` variables in `vars/main.yml`.

### CODE REVIEW

**Comment by UtkarshVerma** [comment]
Fixes #69

File: templates/openwisp2/settings.py [hunk 1]
```diff
@@ -173,6 +174,47 @@
 USE_L10N = True
 USE_TZ = True
 
+#Initialize django-cors-headers
+{% if cors:enable %}
```
Comment (nemesisdesign) [inline templates/openwisp2/settings.py]: look at what we did in the gzip patch and do it in a similar way and add attributes only if present in the dictionary

File: vars/main.yml [hunk 2]
```diff
@@ -1 +1,21 @@
 virtualenv_path: "{{ openwisp2_path }}/env"
+Dictionary for Cross-Origin Resource Sharing (CORS) Settings
+- cors:                                    
+    enable: False                       # Variable to enable/disable django-cors-headers.
```
Comment (nemesisdesign) [inline vars/main.yml]: use `enabled`

File: templates/openwisp2/settings.py [hunk 3]
```diff
@@ -81,6 +82,7 @@
 ]
 
 MIDDLEWARE_CLASSES = [
+    'django.middleware.common.CommonMiddleware',
```
Comment (nemesisdesign) [inline templates/openwisp2/settings.py]: hey, this this already present, why are you adding it here?

File: vars/main.yml [hunk 4]
```diff
@@ -1 +1,21 @@
...
+    methods: [] 
+    headers_defaults: True              #Set to true if defualt headers aren't included in your passed header list.
+    headers: [] 
+    expose_headers: []
+    preflight: 86400
+    credentials: False                         #Set to True to enable cookies in cross-site HTTPS requests
+    model: None
+    replace_referer: False                  
+
+# The usage of django-cors-headers is documented here: "https://pypi.python.org/pypi/django-cors-headers/2.1.0"
```
Comment (nemesisdesign) [inline vars/main.yml]: put this comment before not after

File: templates/openwisp2/settings.py [hunk 5]
```diff
@@ -156,7 +158,6 @@
 
 # Password validation
 # https://docs.djangoproject.com/en/1.9/ref/settings/#auth-password-validators
-
```
Comment (nemesisdesign) [inline templates/openwisp2/settings.py]: never remove lines that are not related to your patch and hence is not necessary to remove them

File: tasks/pip.yml [hunk 6]
```diff
@@ -120,3 +120,12 @@
...
   notify: reload supervisor
+
+- name: Install CORS
+  pip:
+    name: django-cors-headers
+    state: latest
+    virtualenv: "{{ virtualenv_path }}"
+    virtualenv_python: "{{ openwisp2_python }}"
+    virtualenv_site_packages: yes
+  notify: reload supervisor
```
Comment (nemesisdesign) [inline tasks/pip.yml]: use `when` to install this only when needed

File: templates/openwisp2/settings.py (hunk 3 above)
Comment (UtkarshVerma) [reply templates/openwisp2/settings.py]: Oh, didn't notice.

File: templates/openwisp2/settings.py (hunk 1 above)
Comment (UtkarshVerma) [reply templates/openwisp2/settings.py]: @nemesisdesign I looked at it and still have no idea what you are referring to :disappointed: . Please elaborate it.

**Comment by UtkarshVerma** [comment]
Sorry for such a huge mess with the commits. I tried squashing them but it won't work. I might be doing something wrong. :cry:

**Comment by UtkarshVerma** [comment]
Did the requested changes!

File: tasks/pip.yml [hunk 7]
```diff
@@ -120,3 +120,13 @@
     virtualenv_python: "{{ openwisp2_python }}"
     virtualenv_site_packages: yes
   notify: reload supervisor
+
+- name: Install CORS
+  when: "{{ cors:enabled }}"
```
Comment (nemesisdesign) [inline tasks/pip.yml]: this way of accessing dictionary keys doesn't look correct. Have you tested the patch?

Comment (UtkarshVerma) [reply tasks/pip.yml]: @nemesisdesign I had run the playbook with CORS enabled and disabled both, and the server was hosted in both the cases. I don't know how to test CORS so that's still a doubt.

Comment (nemesisdesign) [reply tasks/pip.yml]: it shouuld be `cors.enabled`, same applies for all the other lines

Comment (UtkarshVerma) [reply tasks/pip.yml]: Ok, I will test it tomorrow with the correction again.

**Comment by UtkarshVerma** [comment]
@nemesisdesign Corrected the dictionary accessing commands.

**Comment by nemesisdesign** [comment]
@UtkarshVerma there are conflicts with the master branch (notice how github warns about this), please rebase on master with: `git rebase master`, solve eventual conflicts and update the PR.

If you have never done this you will learn a new thing! :-)

File: templates/openwisp2/settings.py [hunk 8]
```diff
@@ -173,6 +174,47 @@
...
+
+# CORS Allow Methods
+{% if cors.methods_defaults %}                                                                                # If default methods of corsheaders are to be added as well.
+CORS_ALLOW_METHODS = default_methods + tuple({{ cors.methods }}) 
+{% else %}                                                                                                                                # If default methods of corsheaders aren't to be added. 
+CORS_ALLOW_METHODS = {{ cors.methods }}
+{% endif %}
+
+#CORS Allow Headers
+{% if cors.headers_defaults %}                                                                                # If default headers of corsheaders are to be added as well.
```
Comment (nemesisdesign) [inline templates/openwisp2/settings.py]: btw why comments in the code are so far on the right? can you move them closer to the text?

Comment (UtkarshVerma) [reply templates/openwisp2/settings.py]: I will. It's in my aesthetics to move the comments a bit away for neatness. But since you asked, going to align everything sidewards.

Comment (UtkarshVerma) [reply templates/openwisp2/settings.py]: They were misaligned due to edits. Aligning them back now.

**Comment by UtkarshVerma** [comment]
@nemesisdesign I made my fork up-to-date using rebase and now there shouldn't be any merge conflicts. It was a useful thing which I learnt today. :smile:

**Comment by nemesisdesign** [comment]
I need to do some clean up before merging.. it will take me a bit more before I have time to merge it but please leave it open.

**Comment by nemesisdesign** [comment]
@UtkarshVerma in the meanwhile if you could squash the commits into one it would be great and we'd also get a new travis build to check if the patch causes any issue

**Comment by UtkarshVerma** [comment]
@nemesisdesign I'll leave the issue open and will squash the commits.

**Comment by nemesisdesign** [comment]
@UtkarshVerma great, let me know when the squash is done

**Comment by UtkarshVerma** [comment]
@nemesisdesign I synced my fork and squashed the commits.

**Comment by UtkarshVerma** [comment]
Abandoning!

**Comment by nemesisdesign** [comment]
Hey, why did you close this and delete the repository @UtkarshVerma?
The code needed some work and I was leaving it after the GSoC 18 ended so I'd have time to work on it to include it. It takes time sometimes to merge these patches..
